lib/resolver-compat.js

```diff
@@ -0,0 +1,317 @@
+	}
+
+	isPathAllowedForModule(path, mod) {
+		if (!super.isPathAllowed(path)) return false;
+		if (mod && (mod.allowTransitive || path.startsWith(mod.path))) return true;
+		return this.externals.some(regex => regex.test(path));
+	}
+
+	registerModule(mod, filename, path, parent, direct) {
+		const trustedParent = this.trustedMods.get(parent);
+		this.trustedMods.set(mod, {
+			filename,
+			path,
+			paths: this.genLookupPaths(path),
+			allowTransitive: (direct && trustedParent && trustedParent.allowTransitive) || this.externals.some(regex => regex.test(filename))
+		});
+	}
+
+	resolveFull(mod, x, options, ext, direct) {
+		this.currMod = undefined;
+		if (!direct) return super.resolveFull(mod, x, options, ext, false);
+		const trustedMod = this.trustedMods.get(mod);
+		if (!trustedMod || mod.path !== trustedMod.path) return super.resolveFull(mod, x, options, ext, false);
+		const paths = [...mod.paths];
+		if (paths.length === trustedMod.length) {
+			for (let i = 0; i < paths.length; i++) {
+				if (paths[i] !== trustedMod.paths[i]) {
+					return super.resolveFull(mod, x, options, ext, false);
+				}
+			}
+		}
+		const extCopy = Object.assign({__proto__: null}, ext);
+		try {
+			this.currMod = trustedMod;
+			return super.resolveFull(trustedMod, x, undefined, extCopy, true);
+		} finally {
+			this.currMod = undefined;
+		}
+	}
+
+	checkAccess(mod, filename) {
+		const trustedMod = this.trustedMods.get(mod);
+		if ((!trustedMod || trustedMod.filename !== filename) && !this.isPathAllowedForModule(filename, undefined)) {
+			throw new VMError(`Module '${filename}' is not allowed to be required. The path is outside the border!`, 'EDENIED');
+		}
+	}
+
+	loadJS(vm, mod, filename) {
+		filename = this.pathResolve(filename);
+		this.checkAccess(mod, filename);
+		const trustedMod = this.trustedMods.get(mod);
+		const script = this.readScript(filename);
+		vm.run(script, {filename, strict: true, module: mod, wrapper: 'none', dirname: trustedMod ? trustedMod.path : mod.path});
+	}
+
+}
+
+function defaultBuiltinLoader(resolver, vm, id) {
+	const mod = resolver.hostRequire(id);
+	return vm.readonly(mod);
+}
+
+const eventsModules = new WeakMap();
+
+function defaultBuiltinLoaderEvents(resolver, vm, id) {
+	return eventsModules.get(vm);
+}
+
+let cacheBufferScript;
+
+function defaultBuiltinLoaderBuffer(resolver, vm, id) {
+	if (!cacheBufferScript) {
+		cacheBufferScript = new VMScript('return buffer=>({Buffer: buffer});', {__proto__: null, filename: 'buffer.js'});
+	}
+	const makeBuffer = vm.run(cacheBufferScript, {__proto__: null, strict: true, wrapper: 'none'});
+	return makeBuffer(Buffer);
+}
+
+let cacheUtilScript;
+
+function defaultBuiltinLoaderUtil(resolver, vm, id) {
+	if (!cacheUtilScript) {
+		cacheUtilScript = new VMScript(`return function inherits(ctor, superCtor) {
+			ctor.super_ = superCtor;
+			Object.setPrototypeOf(ctor.prototype, superCtor.prototype);
+		}`, {__proto__: null, filename: 'util.js'});
```

Comment:
```js
setPrototypeOf(target, value) {
  // Note: target@this(unsafe) throws@this(unsafe)
  throw new VMError(OPNA);
}
```
I have some questions about `inherits`:
- Why forbidden to use `inherits` in vm2?
- As mentioned in the above code, why not just throw an exception but use `setPrototypeOf` instead

Reply:
This actually allows to use `inherits` in the sandbox. Without it would not work for sandboxed objects and would allow the sandbox to change the prototype chain of host objects passed into the sandbox. The code you copied will only be called when `inherits` is used on a host object where it should not work.